src/test/java/htsjdk/samtools/SAMFileRoundTripTest.java

```diff
@@ -0,0 +1,94 @@
+    public void Utf8RoundTripPositiveTests(final String inputFile, SAMProgramRecord programRecord,final String commentRecord, SAMSequenceRecord sequenceRecord) throws Exception {
+        final File input = new File(TEST_DATA_DIR, inputFile);
+        final File outputFile = File.createTempFile("roundtrip-utf8-out", ".sam");
+        outputFile.delete();
+        outputFile.deleteOnExit();
+        final SAMFileWriterFactory factory = new SAMFileWriterFactory();
+        try (SamReader reader = SamReaderFactory.makeDefault().open(input);
+             SAMFileWriter writer = factory.makeSAMWriter(reader.getFileHeader(), false, new FileOutputStream(outputFile))) {
+            for (SAMRecord rec : reader) {
+                writer.addAlignment(rec);
+            }
+            SAMFileHeader head = reader.getFileHeader();
+            Assert.assertEquals(head.getProgramRecords().get(0), programRecord);
+            Assert.assertEquals(head.getComments().get(0),commentRecord );
+            Assert.assertEquals(head.getSequence("chr3"),sequenceRecord);
+        }
+
+        final String originalsam;
+        try (InputStream is = new FileInputStream(input)) {
+            originalsam = IOUtil.readFully(is);
+        }
+
+        final String writtenSam;
+        try (InputStream is = new FileInputStream(outputFile)) {
+            writtenSam = IOUtil.readFully(is);
+        }
+
+        Assert.assertEquals(writtenSam, originalsam);
+    }
+
+    @DataProvider(name = "Utf8NegativeTestCases")
+    public Object[][] Utf8NegativeTestCases() {
+        return new Object[][]{
+                {"roundtrip_with_utf8_bad_1.sam", "Invalid character in read bases"},
+                {"roundtrip_with_utf8_bad_2.sam", "Non-numeric value in POS column"},
+                {"roundtrip_with_utf8_bad_2.sam", "Non-numeric value in POS column"}
+        };
+    }
+
+    @Test(dataProvider = "Utf8NegativeTestCases",description = "Test UTF-8 encoding present in unpermitted fields of a SAM file", expectedExceptions = {IllegalArgumentException.class, SAMFormatException.class })
+    public void Utf8RoundTripNegativeTest(final String inputFile,final String exceptionString) throws Exception {
+        final File input = new File(TEST_DATA_DIR, inputFile);
+        final File outputFile = File.createTempFile("roundtrip-utf8-out", ".sam");
+        outputFile.delete();
+        outputFile.deleteOnExit();
+        final SAMFileWriterFactory factory = new SAMFileWriterFactory();
+        try (SamReader reader = SamReaderFactory.makeDefault().open(input);
+             SAMFileWriter writer = factory.makeSAMWriter(reader.getFileHeader(), false, new FileOutputStream(outputFile))) {
+            for (SAMRecord rec : reader) {
+                writer.addAlignment(rec);
+            }
+        }
+        catch (final Exception ex) {
+            Assert.assertTrue(ex.getMessage().contains(exceptionString));
+            throw ex;
```

Comment:
Its a bit hard to tell if this test is throwing on read or on write. We should replace this test with separate tests that show that we fail in both cases. Since we will presumably fail when we read a misplaced multibyte character, the test that shows that we fail on write will have to create the output programmatically, rather then via roundtrip.